go/mysql/json/marshal.go

```diff
@@ -149,12 +149,15 @@ func (v *Value) marshalSQLInternal(top bool, dst []byte) []byte {
 		}
 		return dst
 	case TypeNull:
+		// JSON null is a 'null' string literal so we cannot use
+		// CAST(null as JSON) because that (surprisingly?) turns
+		// into an SQL NULL.
 		if top {
-			dst = append(dst, "CAST("...)
+			dst = append(dst, "CONVERT('"...)
```

Comment:
I think this is semantically incorrect. The goal is that JSON marshaling also returns a JSON type from the expression, but it does not for this. It kinda works though because on insert now MySQL casts, but that's an implementation side-effect for `vreplication`, not something generic.

I think this should be `CAST('null' as JSON)` (instead of `CAST(null as JSON)`), which should also work I think and returns a JSON type for the expression:

```
mysql> select CAST(null as JSON);
+--------------------+
| CAST(null as JSON) |
+--------------------+
| NULL               |
+--------------------+
1 row in set (0.00 sec)

mysql> select CAST('null' as JSON);
+----------------------+
| CAST('null' as JSON) |
+----------------------+
| null                 |
+----------------------+
1 row in set (0.00 sec)
```

Given the MySQL documentation, I think we also should change the serialization for `true` / `false` as a toplevel value, and also quote that with a single `'` to match the behavior for how we implement `null`.

Reply:
@dbussink I agree. And that's what I tried first, but then it became clear that there was another issue (and CONVERT is what we were previously using here):
```
+----------+-------------+-------+-------+--------------------------+
| order_id | customer_id | sku   | price | json_data                |
+----------+-------------+-------+-------+--------------------------+
|        1 |           1 | sku1  |     1 | NULL                     |
|        2 |           2 | sku2  |     2 | {}                       |
|        3 |           3 | sku3  |     3 | "base64:type15:bnVsbA==" |
|        4 |           4 | sku4  |     4 | true                     |
|        5 |           5 | sku5  |     5 | false                    |
|        6 |          10 | sku10 |    10 | NULL                     |
|        7 |           6 | sku6  |     6 | true                     |
|        8 |           7 | sku7  |     7 | false                    |
|        9 |           8 | sku8  |     8 | "base64:type15:bnVsbA==" |
|       10 |           9 | sku9  |     9 | NULL                     |
+----------+-------------+-------+-------+--------------------------+
```

This seems to lead us to treating it subsequently as a binary blob. If you think that's another internal bug then I can track that down too. And if you might have any ideas... 🙂 

Thanks!

Reply:
Same thing for 'true' and 'false' as well — the issue I noted above is not specific to 'null'. I had initially tried to treat true/false/nil the same way (as you suggested above).

Reply:
Does that mean it needs `CAST(CONVERT('null' using utf8mb4) as JSON)`, or syntactically different but semantically the same, `CAST(_utf8mb4'null' as JSON)`. I know that's messy, but I think that's what needed to be correct.

That it's otherwise binary is due to vreplication using the `binary` collation and we need to make sure then that inputs are encoding agnostic / explicit about it.

Reply:
Thanks, @dbussink ! That makes sense. I can see that we use the `_utf8mb4'literal'` syntax throughout our JSON handling so I switched to that and it's working as expected.